beaker_kernel/lib/subkernels/python.py

```diff
@@ -52,3 +55,51 @@ def parse_subkernel_return(cls, execution_result) -> Any:
         if return_str:
             python_obj = ast.literal_eval(return_str)
             return python_obj
+
+    async def generate_checkpoint_from_state(self) -> Checkpoint:
+        save_state_code = """
+import inspect as _inspect
+import json as _json
+import dill as _dill
+class _SubkernelStateEncoder(_json.JSONEncoder):
+    def default(self, o):
+        # if callable(o):
+            # return f"Function named"
+            # return super().default(o)
+        try:
+            return super().default(o)
+        except:
+            return str(o)
+
+_result = {}
+for _name, _value in dict(locals()).items():
```

Comment:
I'd be curious whether this will work as we'd expect with xarray since xarray datasets may be treated like open file handles under the hood. Would be worth testing and also thinking through possible other cases this might not work as expected

Reply:
This is a good edge case I need to try

Reply:
I tested xarray on the code below and it worked:
```
import xarray as xr
import numpy as np
import pandas as pd

time = pd.date_range('2024-01-01', periods=100, freq='D')
lat = np.linspace(-90, 90, 180)
lon = np.linspace(-180, 180, 360)
temperature = np.random.rand(100, 180, 360) * 20 + 15
temp_da = xr.DataArray(data=temperature,
                       dims=["time", "latitude", "longitude"],
                       coords={"time": time, "latitude": lat, "longitude": lon},
                       name='temperature')
```

after I ran 
- save
- `temp_da = None`
- undo

Reply:
I also tried the same thing with `ds = temp_da.to_dataset()` and it worked.